### ISSUE TITLE
Justified gallery (ala Flickr or Google+)

### ISSUE THREAD

**Issue by msharp1** [thumbsup/thumbsup#86]
Hi,
I would like to see a justified/linear gallery theme for Thumbsup.
Why not using [the Justified Gallery](http://miromannino.github.io/Justified-Gallery/) JQuery module. What do you think ?

Regards,

**Comment by rprieto** [comment]
Hi @msharp1, that type of view looks very nice!

I can think of 1 main blocker for the moment: I assume the reason you like it is because is mixes the the different-sized thumbnails in a very nice way? For now, to avoid this problem `thumbsup` always generates square thumbnails - similar to the iOS Photos app.

I'd need to have a think, does it make more sense to:
- always generate original-shape thumbnails, and let the themes deal with it?
- generate both original-shape and square thumbnails, and let the theme choose which one to use?
- have an option to choose which thumbnail to generate (to speed up the build), knowing that some themes will become incompatible?

Once we've solved that, using the Justified Gallery plugin would be great - either on an existing theme or a new one. What do you think?

**Comment by msharp1** [comment]
Hi @rprieto,
Sorry for this late answer and thank you for answering so fast on your side...
I think I would be for the first option because the "justified" display isn't the only one that could need original shape thumbs in the future.
I am also in favor not adding a configuration option unless really needed, so you keep the software more straight forward to use and install.
A question by the way. Let's assume you have the original shape thumb. Is it possible to generate a cropped version of it without re-compressing ? That could be another solution. I suppose it is also related to the input format.

**Comment by amcolash** [comment]
Love the look of that gallery. I'd assume that using rectangular thumbs (consistent height) and then using `cover` for the css of a background image would mimic that behavior. If I get some time I may take a stab at this, since I like the look slot. The one issue is with a very tall image and things look off since it is so narrow

**Comment by piccolbo** [comment]
Very tall images are not very common, whereas panoramas are a mainstream type of photography. Your observation is correct but it happens so occasionally that it may be a case of the perfect getting in the way of the good.

**Comment by amcolash** [comment]
On the whole, I am liking how this looks with portrait-orientated photos. I needed to change the structure of the generated pages a bit which @rprieto might have feedback on (changed `li`s to `div`s), but otherwise looks nice with the cards theme!

![image](https://user-images.githubusercontent.com/8282413/52542654-f6a5a900-2d56-11e9-818c-e63a27dc5034.png)

 I am tempted to say that we should make this a new theme since it is pretty different from the base themes so far. That being said, the way thumbnails are now generated would still require an update to the existing themes (and may change how they look a bit).

**Comment by rprieto** [comment]
Thanks @amcolash, the result looks great! At first I though it could be a separate theme, but I think it fits the Cards theme really well. I'll be happy to merge a PR over there.

What media did you use to generate the different shapes? The square thumbs with CSS "cover"? Or the large shapes so the result is sharper?

To give the themes more flexibility I wonder if thumbsup should generate:
- a `100x100` thumbnail
- a `300x_` small image **<<< new**
- a `1000x_` large image

Otherwise if you use 100x thumbs for a small theme, you won't be able to change for a theme that needs bigger images without rebuilding the whole gallery. What do you think?

**Comment by amcolash** [comment]
I did `?x250` for the sizes. In the resize step I just took out the width and it resized to keep heights consistent. I do like the slightly larger images for thumbnails...

I would argue that changing all thumbnails to  being actually correct on the aspect ratio would be ideal (though super tall images might get broken, bit this could be solved with more complex logic). This gives a consistent theme development experience and those who want squares can use the `cover` method (and less resources for less copies of things).

The regenerate issue is interesting. I tried on a test set of files before and after the size change. It looked like I could just wipe the thumbnail folder and it did not need to resize the large ones. Would an upgrade path be worth it here?

Crazy idea, what if the theme provided a list of image sizes it expected? Then it is totally up to the theme to decide what size thumbnails are - though much more dev cost with little benefit. I suppose the user can still override that value as an argument anyways...

Otherwise it could be that a theme specifies expected image buckets, i.e. thumbnail_cropped, thumbnail_uncropped, large, etc and then only relevant files are generated. That makes upgrade easy and keeps overhead lower for those who don't use the theme. We could even purge old thumbnails based on theme requirements to save space.

**Comment by amcolash** [comment]
After some more investigation, it appears that the pager plugin for light gallery is not too happy with non-square thumbnails. It may make sense to just go with both versions of the thumbnail because of this. I do still like the idea of a theme "opting in" to have the other resolution images if people do not need 2 sets of thumbnails.

**Comment by rprieto** [comment]
Thanks for all your investigations! I think it would be tricky for themes to specify the exact sizes they need (e.g. `250px`) because changing theme would often mean re-building all thumbnails. However I like the idea that themes could choose between a few presets.

For now they could *always* be generated, and in the future the theme could give a hint whether it needs them. We'd have to think about it some more, because it could also be a strange experience if
- you use Theme A which builds all square thumbnails
- you switch to Theme B which deletes them all
- you switch back to Theme A and have to wait 10min to rebuild them again (I run galleries with 10k pictures so always try to keep in mind how a feature will scale).

Regarding your last comment above: if we generated **both** square thumbnails and proportional thumbnails, would that help Lightgallery? For example could you have the nice "justified gallery" page, and still use the square thumbnails for the lightgallery previews? (it probably depends on the markup we use).

**Comment by amcolash** [comment]
I think I have come up with a reasonable solution. It adds some options that themes can specify in their `package.json`. I will make a PR and we can continue the discussion there. It will only generate extra files if needed but otherwise no changes for existing themes (until cards is updated).

**Comment by rprieto** [comment]
Thank you, I'll have to think about it some more. Themes specifying sizes and what to generate is what I mentioned I wanted to avoid in the last message.

Do you have any objections to always generating the rectangular thumbnail, say with a default size of 250? This would allow users to change theme without rebuilding any files. I think if we choose sizes wisely (like `thumb=100, small=250, large=1000`) it should give themes a lot of flexibility in how to display data. It reminds me of how Synology implements their PhotoStation product:

```
64212 Jan 16 10:36 SYNOPHOTO:THUMB_B.jpg
-rwxrwxrwx    1 root     root        149043 Jan 16 10:36 SYNOPHOTO:THUMB_L.jpg
-rwxrwxrwx    1 root     root         25034 Jan 16 10:36 SYNOPHOTO:THUMB_M.jpg
-rwxrwxrwx    1 root     root          3807 Jan 16 10:36 SYNOPHOTO:THUMB_S.jpg
-rwxrwxrwx    1 root     root        328279 Jan 16 10:36 SYNOPHOTO:THUMB_XL.jpg
```

**Comment by rprieto** [comment]
Did you have a change to see if lightgallery would work if we have both the square thumbnails and the small resized image?

**Comment by amcolash** [comment]
At the moment it is doing exactly that. If you opt-in to the rectangular size, it will add those to the list to generate. If not, it will not generate those files. The dimensions are an opt-in so that themes that need it can push those sizes larger (that way a theme can customize how large the thumbnails are if it needs), but otherwise it will default to the width of a normal thumbnail. I am fine with removing the sizes if we think 250 is sufficient, but I would like to keep that opt-in theme by theme behavior (since not all themes may need those extra thumbnails).

**Comment by rprieto** [comment]
I'm afraid that if we allow themes to specify `defaultThumbnailSize`, we might end up with each theme customising exactly what they need (200, 230, 300). This would break the principle of least surprise because:
- when you switch themes, `thumbsup` won't know the files need to be re-generated, so the theme will expect a 300px image but might actually get 150px
- we could tell users to always delete all thumbnails when switching themes, but for large galleries that's a pretty big ask

I had something much simpler in mind: always generate a `small` resized image, say 250px. The themes can then choose between `url.thumb` (tiny square), `url.small` (small rectangle) and `url.large` (large rectangle) depending on what they want to use. This model seems tried and tested in [PhotoStation](https://www.synology.com/fr-fr/dsm/feature/photo_station) and [Sigal](http://sigal.saimon.org/en/latest/configuration.html). I think it allows users to switch themes easily, and keeps theme-development very standard. I would even consider removing the `--thumb-size` option to keep things simple.

We can also see it as a simple first step in the right direction - and still have the option to make it theme-configurable in the future.

**Comment by rprieto** [comment]
Update: I checked my largest gallery which is 100GB of input data, and the `thumbs` folder is 100MB. Even if "small" output files are a little heavier, I don't think people will mind if we always generate them and it adds ~0.5% of the input size.

I'd say it's a way to keep it simple until we figure out if there's demand for turning it on/off.

**Comment by amcolash** [comment]
Those are fair points for sure. I do prefer the idea of opting in to the small thumbnails but totally get your concerns on the complexity of such a system. For now, I can yank those options from the themes and just keep it simple. Since that 100GB translates to 100MB, I would assume we are looking at about 200-300MB of equivalent small thumbs.

Thanks for all of the input on this, I am excited to see it in action on my own server soon :)

**Comment by rprieto** [comment]
Thanks for your idea and being flexible too! This is a great step towards more & better themes. I'll close this to focus on your PR.

**Comment by rprieto** [comment]
Thanks everyone for your ideas! This is a great step towards more & better themes. I'll close this to focus on @amcolash's PR.

**Comment by piccolbo** [comment]
If the PR that you reference is #146, that PR says it is "part of the solution to #86" (this ticket), not equivalent to it. It is now merged and I don't see any changes to a UI or documentation so  that I could start using it, which seems to confirm it is not a complete solution to #86. Could you please clarify? Thanks!

**Comment by rprieto** [comment]
Sorry maybe I closed it a little early. The changes are on master, and allow any theme to create justified galleries. A new version will be published soon and the docs updated when published. The docs will only change in the "themes" section since nothing else changes from a user point of view.

The Cards theme will be updated to be justified, unless there's consensus it should stay as-is for the people who are using it.

**Comment by piccolbo** [comment]
I gave it a try and it does work. I have some ideas for a theme based on it that I'll report elsewhere.

**Comment by amcolash** [comment]
Since the cards theme PR (https://github.com/thumbsup/theme-cards/pull/3) is not yet finalized, I would check there @piccolbo.

**Comment by rprieto** [comment]
Just an update that rectangular thumbnails are available in version `2.9.0`! I will add `--theme flow` as soon as a version of https://github.com/thumbsup/theme-flow is published to npm.
